### PR TITLE
Bugfix/issue 425 multishard incorrect result on distinct

### LINKED ISSUE #425: multishard incorrect result on aggr_distinct (with append only stream)
**Describe what's wrong**
```
client@1:
create stream v2(id int, val int) primary key id settings shards=3;
client@1:
select sum_distinct(val), sum(val) from v2
client@2:
insert into v2(id, val) values(3, 30), (4, 40)
client@2:
insert into v2(id, val) values(1, 30), (7, 40)
```

**Error message and/or stacktrace**

```
c1 :) select sum_distinct(val), sum(val) from v2

SELECT
  sum_distinct(val), sum(val)
FROM
  v2

Query id: 695de04b-70bf-4d6e-9497-319d339d8c43

┌─sum_distinct_streaming(val)─┬─sum(val)─┐
│                          70 │       70 │
└─────────────────────────────┴──────────┘
┌─sum_distinct_streaming(val)─┬─sum(val)─┐
│                         100 │      140 │
```
**Additional context**

the result shall be 70 right?

### PR DESCRIPTION
**Pull request by lizhou1111**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:
this close #425

### CODE REVIEW

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 1]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
     std::vector<T> extra_data_since_last_finalize;
-    bool use_extra_data = false;  /// Optimized, only streaming global aggreagtion query need to use extra data after first finalization.
-    /// proton: ends.
+
+    /// If has new data
+    bool has_new_data = false;
+
+    /// A blank mark that's used to decide which kind is the current block. 
+    /// 0 means data block, 1 means a special data block whose extra data has been deleted, 2 means a blank block.
+    int is_blank = 0;
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: we can use `enum class` instead, and add examples to explain these values

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 2]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
 
     void add(const IColumn ** columns, size_t /* columns_num */, size_t row_num, Arena *)
     {
         const auto & vec = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
-        /// proton: starts.
         auto [_, inserted] = set.insert(vec[row_num]);
-        if (use_extra_data && inserted)
+        has_new_data = false;
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: if only insert existed values ,  the flag `has_new_data` will be set false , is right ?

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 3]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
+        if (inserted)
+        {
+            has_new_data = true;
             extra_data_since_last_finalize.emplace_back(vec[row_num]);
-        /// proton: ends.
+        }
+        if (is_blank == 2)
+        {
+            is_blank = 1;
+        }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: 2 -> 1 , what's mean ?

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 4]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
-                auto [_, inserted] = set.insert(elem.getValue());
-                if (inserted)
-                    extra_data_since_last_finalize.emplace_back(elem.getValue());
+                for (const auto & data : rhs.extra_data_since_last_finalize)
+                {
+                    auto [_, inserted] = set.insert(data);
+                    if (inserted)
+                    {
+                        extra_data_since_last_finalize.emplace_back(data);
+                    }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
if (inserted)
    extra_data_since_last_finalize.emplace_back(data);
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 5]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
         {
-            for (const auto & elem : rhs.set)
+            for (const auto & data : rhs.set)
+            {
+                double value = data.getValue();
+                auto it = std::find(extra_data_since_last_finalize.begin(), extra_data_since_last_finalize.end(), value);
+                if (it != extra_data_since_last_finalize.end())
+                {
+                    extra_data_since_last_finalize.erase(it);
+                }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h:66]:
```suggestion
if (it != extra_data_since_last_finalize.end())
    extra_data_since_last_finalize.erase(it);
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 6]
```diff
@@ -19,84 +19,130 @@ struct AggregateFunctionDistinctSingleNumericData
...
+            is_blank = 2;
         }
-        else if (use_extra_data)
+        /// if the block is in 1 mode, merge the set and delete the data both existed in the current extra data and the set.
+        if (rhs.is_blank == 1)
         {
-            for (const auto & elem : rhs.set)
+            for (const auto & data : rhs.set)
+            {
+                double value = data.getValue();
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: is double type ?

Comment (yl-lisen) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: should be `T`

Comment (yl-lisen) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h:61]: why dose need call `getValue()`

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h (hunk 5 above)
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: remove `{ }` for single line

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h (hunk 6 above)
Comment (yl-lisen) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
                auto value = data.getValue();
```

File: tests/stream/test_stream_smoke/0012_multishards7.yaml [hunk 7]
```diff
@@ -452,3 +452,161 @@ tests:
...
+            query_type: table
+            wait: 3
+            query: |
+              drop stream if exists test13_multishard_7;
+    
+    expected_results:
+      - query_id: '1366'
+        expected_results:
+          - [70, 70]
+          - [770, 770]
```
Comment (yl-lisen) [inline tests/stream/test_stream_smoke/0012_multishards7.yaml]: add blank line at end of file

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h (hunk 2 above)
Comment (Jasmine-ge) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: Yes, the has_new_data only works when there is new data been inserted

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 8]
```diff
@@ -10,7 +10,16 @@ namespace DB
...
+enum class IsBlank
+{
+    /// A block that contains new data, whose data will not be deleted.
+    DataBlock,
+    /// A block that appears in special occasion, as the first insertion didn't call merge, the extra_data of this block will be deleted.
+    /// So we'll need to mark the block and merge all the data in set to check if there's an overlap
+    LessExtraDataBlock,
+    /// A block that is newly created for merge, its extra_data must have been cleared.
+    BlankBlock,
+};
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: remove it

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 9]
```diff
@@ -19,84 +28,77 @@ struct AggregateFunctionDistinctSingleNumericData
...
     {
         const auto & vec = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
-        /// proton: starts.
         auto [_, inserted] = set.insert(vec[row_num]);
-        if (use_extra_data && inserted)
+        if (inserted)
+        {
             extra_data_since_last_finalize.emplace_back(vec[row_num]);
-        /// proton: ends.
+        }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
        if (inserted)
            extra_data_since_last_finalize.emplace_back(vec[row_num]);
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 10]
```diff
@@ -19,84 +28,77 @@ struct AggregateFunctionDistinctSingleNumericData
...
     }
 
-    void merge(const Self & rhs, Arena *)
+    void merge(const Self & rhs, Arena *, ConstAggregateDataPtr __restrict place)
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: don't need pass in `ConstAggregateDataPtr __restrict place`,
we can use `reinterpret_cast<uintptr_t>(&rhs)` instead

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 11]
```diff
@@ -19,84 +28,77 @@ struct AggregateFunctionDistinctSingleNumericData
...
-        if (rhs.use_extra_data)
+        for (const auto & data : extra_data_since_last_finalize)
         {
-            for (const auto & data : rhs.extra_data_since_last_finalize)
-            {
-                auto [_, inserted] = set.insert(data);
-                if (use_extra_data && inserted)
-                    extra_data_since_last_finalize.emplace_back(data);
-            }
+            if ((rhs.set).find(data)){
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: have a format

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 12]
```diff
@@ -19,84 +28,77 @@ struct AggregateFunctionDistinctSingleNumericData
...
+                auto it = std::find(extra_data_since_last_finalize.begin(), extra_data_since_last_finalize.end(), data);
+                if (it != extra_data_since_last_finalize.end())
+                {
+                    extra_data_since_last_finalize.erase(it);
+                } 
+            }        
         }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
        /// Deduplicate owned extra data based on rhs
        for (auto it = extra_data_since_last_finalize.begin(); it != extra_data_since_last_finalize.end();)
        {
            if (rhs.set.find(*it) != rhs.set.end())
                it  = extra_data_since_last_finalize.begin.erase(it);
            else
                ++it;
        }
        
        /// Merge and deduplicate rhs' extra data
        ... similar behavior
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 13]
```diff
@@ -19,84 +28,77 @@ struct AggregateFunctionDistinctSingleNumericData
...
-                if (inserted)
-                    extra_data_since_last_finalize.emplace_back(elem.getValue());
-            }
+            auto [_, inserted] = set.insert(data);
+            if (inserted)
+                extra_data_since_last_finalize.emplace_back(data);
         }
-        else
+    
+        set.merge(rhs.set);
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h:123]: don't need merge set ? we have merged extra data

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 14]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
+        for (auto it = extra_data_since_last_finalize.begin(); it != extra_data_since_last_finalize.end();)
         {
-            for (const auto & data : rhs.extra_data_since_last_finalize)
-            {
-                auto [_, inserted] = set.insert(data);
-                if (use_extra_data && inserted)
-                    extra_data_since_last_finalize.emplace_back(data);
-            }
+            if (rhs.set.find(*it) != rhs.set.end())
+                extra_data_since_last_finalize.erase(it);
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
                it = extra_data_since_last_finalize.erase(it);
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 15]
```diff
@@ -108,27 +99,43 @@ struct AggregateFunctionDistinctGenericData
...
         Set::LookupResult it;
         bool inserted;
-        for (const auto & elem : rhs.set)
-        /// proton: starts.
+        /// Deduplicate owned extra data based on rhs
+        for (auto next = extra_data_since_last_finalize.begin(); next != extra_data_since_last_finalize.end();)
         {
-            set.emplace(ArenaKeyHolder{elem.getValue(), *arena}, it, inserted);
+            if (rhs.set.find(*next) != rhs.set.end())
+                extra_data_since_last_finalize.erase(next);
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
                next = extra_data_since_last_finalize.erase(next);
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 16]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
-            }
+            auto [_, inserted] = set.insert(data);
+            if (inserted)
+                extra_data_since_last_finalize.emplace_back(data);
         }
-        else
+
+        set.merge(rhs.set);
+
+        if ((rhs.extra_data_since_last_finalize).size())
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h:59]:
```suggestion
        if (!rhs.extra_data_since_last_finalize.empty())
```

Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: we don't need this condition, please save all merged places

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 17]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
         {
-            set.merge(rhs.set);
+            uintptr_t temp = reinterpret_cast<uintptr_t>(&rhs);
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: `temp` -> `merged_place`

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 18]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
+            auto find_place = std::find(related_places.begin(), related_places.end(),temp);
+            if (find_place == related_places.end())
+                related_places.emplace_back(temp);
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: `related_places` -> `merged_places`

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 19]
```diff
@@ -108,27 +99,43 @@ struct AggregateFunctionDistinctGenericData
...
     using Self = AggregateFunctionDistinctGenericData;
     Set set;
-    /// proton: starts. Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Optimized, put the new coming data that the set does not have into extra_data_since_last_finalize.
     std::vector<StringRef> extra_data_since_last_finalize;
-    bool use_extra_data = false;  /// Optimized, only streaming global aggreagtion query need to use extra data after first finalization.
-    /// proton: ends.
-
+    std::vector<uintptr_t> related_places;
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
    NO_SERDE std::vector<uintptr_t> related_places;
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 20]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
     Set set;
 
-    /// proton: starts. Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Optimized, put the new coming data that the set does not have into extra_data_since_last_finalize.
     std::vector<T> extra_data_since_last_finalize;
-    bool use_extra_data = false;  /// Optimized, only streaming global aggreagtion query need to use extra data after first finalization.
-    /// proton: ends.
 
+    std::vector<uintptr_t> related_places;
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
    NO_SERDE std::vector<uintptr_t> merged_places;
```

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 21]
```diff
@@ -333,10 +308,26 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
...
 
-        /// proton: starts. Next finalization will use extra data, used in streaming global aggregation query.
-        this->data(place).use_extra_data = true;
-        this->data(place).extra_data_since_last_finalize.clear();
-        /// proton: ends.
+
+        /// clear all the extra data in related blocks
+        for (auto & item : this->data(place).related_places) {
+            this->data(reinterpret_cast<AggregateDataPtr>(item)).extra_data_since_last_finalize.clear();
+        }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: remove `{}`

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 22]
```diff
@@ -333,10 +308,26 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
...
+
+        /// Add the temp data to sum so that the distinct outcome can accumulate
+        if(this->data(place).related_places.size()){
+            nested_func->addBatchSinglePlace(0, arguments[0]->size(), getNestedPlace(reinterpret_cast<AggregateDataPtr>(this->data(place).related_places[0])), arguments_raw.data(), arena);
+        }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: remove {}

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 23]
```diff
@@ -333,10 +308,26 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
...
+
+        if (this->data(place).extra_data_since_last_finalize.size())
+        {
+            this->data(place).extra_data_since_last_finalize.clear();
+        }
+
+        if (this->data(place).related_places.size())
+        {
+            this->data(place).related_places.clear();
+        }
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]:
```suggestion
        this->data(place).extra_data_since_last_finalize.clear();
        this->data(place).related_places.clear();  
```

Comment (yl-lisen) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: add a whole example to explain this logic

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 24]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
+    /// Optimized, put the new coming data that the set does not have into extra_data_since_last_finalize.
     std::vector<T> extra_data_since_last_finalize;
-    bool use_extra_data = false;  /// Optimized, only streaming global aggreagtion query need to use extra data after first finalization.
-    /// proton: ends.
 
+    std::vector<uintptr_t> related_places;
+
+    /// not used
+    bool use_extra_data = false;
     void add(const IColumn ** columns, size_t /* columns_num */, size_t row_num, Arena *)
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: Add a blank line between

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 25]
```diff
@@ -19,84 +18,76 @@ struct AggregateFunctionDistinctSingleNumericData
...
-    /// proton: starts. Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Resolve multiple finalizations problem for streaming global aggreagtion query
+    /// Optimized, put the new coming data that the set does not have into extra_data_since_last_finalize.
     std::vector<T> extra_data_since_last_finalize;
-    bool use_extra_data = false;  /// Optimized, only streaming global aggreagtion query need to use extra data after first finalization.
-    /// proton: ends.
 
+    std::vector<uintptr_t> related_places;
+
+    /// not used
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: `It's not used in new implementation, but is kept for backward compatibility`

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 26]
```diff
@@ -5,12 +5,11 @@
 #include <Columns/ColumnArray.h>
 #include <Common/assert_cast.h>
 #include <Common/HashTable/HashSet.h>
-
+#include <Common/serde.h>
 namespace DB
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: add a blank line between

Comment (yl-lisen) [reply src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: need have a format for this file

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 27]
```diff
@@ -19,84 +19,73 @@ struct AggregateFunctionDistinctSingleNumericData
...
     void add(const IColumn ** columns, size_t /* columns_num */, size_t row_num, Arena *)
     {
         const auto & vec = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
-        /// proton: starts.
         auto [_, inserted] = set.insert(vec[row_num]);
-        if (use_extra_data && inserted)
+        if (inserted)
             extra_data_since_last_finalize.emplace_back(vec[row_num]);
-        /// proton: ends.
+
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: remove this line

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 28]
```diff
@@ -327,16 +302,33 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
...
+         */
+        if(this->data(place).merged_places.size())
+            nested_func->addBatchSinglePlace(0, arguments[0]->size(), getNestedPlace(reinterpret_cast<AggregateDataPtr>(this->data(place).merged_places[0])), arguments_raw.data(), arena);
+
+        if (this->data(place).extra_data_since_last_finalize.size())
+            this->data(place).extra_data_since_last_finalize.clear();
+
+        if (this->data(place).merged_places.size())
+            this->data(place).merged_places.clear();
+
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: No need to check if they exist, we can directly clean them up

File: src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h [hunk 29]
```diff
@@ -327,16 +302,33 @@ class AggregateFunctionDistinct : public IAggregateFunctionDataHelper<Data, Aggr
...
+        /// Clear all the extra data in related blocks
+        for (auto & item : this->data(place).merged_places)
+            this->data(reinterpret_cast<AggregateDataPtr>(item)).extra_data_since_last_finalize.clear();
+
+        /* Add the temp data to sum so that the distinct outcome can accumulate.
+         The block we using here may be a blank block, therefore we need to find data block and store sum in it.
+         For example: if the current sum is 70, what we stored in sum{} of blank block will be 70, but it will be
+         deleted, so we find the information of the first block stored in merged_places and put 70 to its sum
+         using addBatchSinglePlace. 
+         */
```
Comment (yl-lisen) [inline src/AggregateFunctions/Streaming/AggregateFunctionDistinct.h]: need a clearer example here or at the begin of file

**Review by yl-lisen — APPROVED** [review]

**Pull request merged**